src/membership/cut_detector.rs

```diff
@@ -0,0 +1,162 @@
+use crate::{
+    common::Endpoint,
+    transport::proto::{AlertMessage, EdgeStatus},
+};
+
+use std::collection::{HashMap, HashSet};
+
+const NUM_MIN: usize = 3;
+
+pub struct MultiNodeCutDetector {
+    /// Number of observers per subject and vice versa
+    num: usize,
+    /// High water mark
+    high: usize,
+    /// Low water mark
+    low: usize,
+    proposal_count: usize,
+    updates_in_progress: usize,
+    reports_per_host: HashMap<Endpoint, HashMap<usize, Endpoint>>,
+    proposal: Vec<Endpoint>,
+    pre_proposal: HashSet<Endpoint>,
+    seen_down_link_events: bool,
+}
+
+impl MultiNodeCutDetector {
+    pub fn new(num: usize, high: usize, low: usize) -> Self {
+        if high > num || low > high || num < NUM_MIN || low <= 0 || high <= 0 {
+            panic!("Invalid arguments passed to cut detector")
```

Comment:
this is fine for now but prob want to use a builder? or asserts? we can punt on this jus thinking.

Reply:
yeah, this can just be an assert. And honestly, I don't think a builder would be helpful. It's an internal API and has 3 constant value being passed to it. Don't see the advantage of having a builder for this.,